### PR TITLE
PhoneUsageProvider

### PR DESCRIPTION
**Pull request by MaximMoinat**
Added Provider for usage events, monitoring all app starts and close events.
This provider also listens for interaction state, including shutdown events.

### CODE REVIEW

File: src/main/java/org/radarcns/phone/PhoneUsageProvider.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public List<String> needsPermissions() {
+        // TODO: add PACKAGE_USAGE_STATS
+        //Check if permission enabled
+//        if (appUsages.getUsageStatsList().isEmpty()){
+//            Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
+//            startActivity(intent);
+//        }
+        return Arrays.asList(WRITE_EXTERNAL_STORAGE, INTERNET);
```
Comment (blootsvoets) [inline src/main/java/org/radarcns/phone/PhoneUsageProvider.java]: `WRITE_EXTERNAL_STORAGE` is not used here, `INTERNET` is already provided by default. Just put `PACKAGE_USAGE_STATS` here, and the app will request the permission.

File: src/main/java/org/radarcns/phone/PhoneUsageManager.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            usageEvent = new UsageEvents.Event();
+        }
+
+        // If no event was found, then the permission needs to be given
+        if (previousEventPackageName == null){
+//            Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
+//            startActivity(intent);
+            logger.info("Please give this app Android Usage Access permissions (Security->App Usage access)");
+            Boast.makeText(context, "Please give this app Android Usage Access permissions (Security->App Usage access)", Toast.LENGTH_LONG).show();
+        }
```
Comment (blootsvoets) [inline src/main/java/org/radarcns/phone/PhoneUsageManager.java]: Should not be needed, use the provider `needsPermissions()` instead (see comment below).

File: src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -8,9 +8,13 @@
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
 
+    <uses-permission android:name="android.permission.PACKAGE_USAGE_STATS"/>
+    <uses-permission android:name="android.permission.INTERNET"/>
```
Comment (blootsvoets) [inline src/main/AndroidManifest.xml]: RADAR-Android-Commons already states `INTERNET` permission, no need to repeat it. `INTERNET` permission is not used here, is it?

**Comment by MaximMoinat** [comment]
Doubted about the receiver for boot action. This moment is also registered as a screen interaction event (assuming user unlocks phone after boot).
What is the added value of listening for boot action?

**Comment by blootsvoets** [comment]
Just to have it explicit, that's about it.

File: src/main/java/org/radarcns/phone/PhoneUsageManager.java [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+        // Get alarm manager and schedule it to run every period (seconds)
+        AlarmManager alarmManager = (AlarmManager) this.context.getSystemService(ALARM_SERVICE);
+        alarmManager.setInexactRepeating(AlarmManager.RTC_WAKEUP, System.currentTimeMillis(), period * 1000, pendingIntent);
+
+        // Activity to perform when alarm is triggered
+        this.context.registerReceiver(new BroadcastReceiver() {
+            @Override
+            public void onReceive(Context context, Intent intent) {
+                processUsageEvents();
```
Comment (blootsvoets) [inline src/main/java/org/radarcns/phone/PhoneUsageManager.java]: Have you tested this call while the app is not running? I'm wondering if the service will get started to actually process the message, or that you'd get errors that the current context is not valid.

File: src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -8,9 +8,13 @@
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
 
+    <uses-permission android:name="android.permission.PACKAGE_USAGE_STATS"/>
+    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
```
Comment (blootsvoets) [inline src/main/AndroidManifest.xml]: Is this permission still needed?

File: src/main/java/org/radarcns/phone/PhoneUsageManager.java (hunk 4 above)
Comment (blootsvoets) [inline src/main/java/org/radarcns/phone/PhoneUsageManager.java]: Ok never mind.

**Pull request merged**